### PR TITLE
New board, SD card, restructuring (update).

### PR DESCRIPTION
**Pull request by elmo128**
Added ESP32cam board (no display implemented, ESP32-D0WD-V3 based), setup via config file on sd card, moved certain parts of code, eg. some of the nvMem access to a dedicated class. As this is a rebased version of an earlier pull request, this class does not yet handle logging and other new features, except enable/disable.

### CODE REVIEW

File: src/mining.cpp [hunk 1]
```diff
@@ -28,11 +29,12 @@ uint32_t valids; // increased if blockhash <= target
...
-extern char poolString[80];
-extern int portNumber;
-extern char btcString[80];
+extern TSettings Settings;
+
 IPAddress serverIP(1, 1, 1, 1); //Temporally save poolIPaddres
-extern bool saveStatsToNVS; //Track mining stats in non volatile memory
+
+//Track mining stats in non volatile memory
+extern TSettings Settings;
```
Comment (romanmashta) [inline src/mining.cpp:37]: Duplicate line with line number 32

Comment (elmo128) [reply src/mining.cpp]: thanks, fixed

**Review by romanmashta — APPROVED** [review]
Nice 👍

File: src/wManager.cpp
Comment (BitMaker-hub) [inline src/wManager.cpp:1]: This should be carefully tested. I'll test on clean boards to see if it properly works or not

**Review by BitMaker-hub — APPROVED** [review]
Ok, checked and approved

**Comment by BitMaker-hub** [comment]
After this merge we will need to recheck all boards.
But current SPIFF config is not compatible with NVME structure you proposed. 
I had to reconfig board.

Can you check if is possible to setup the NVME partition on same place were current json config file is?

**Comment by BitMaker-hub** [comment]
Sorry, that was not part of your job. This was merged before, but some of your changes is chrashing ESP32-2432S028R board

**Comment by elmo128** [comment]
> After this merge we will need to recheck all boards. But current SPIFF config is not compatible with NVME structure you proposed. I had to reconfig board.

The config files on flash and SD card share the same set of keys. I changed them to be more self-explanatory, as it is intended for users to modify their value on the card.

> Can you check if is possible to setup the NVME partition on same place were current json config file is?

It should still be in the same location, only the keys differ.

**Comment by elmo128** [comment]
> Sorry, that was not part of your job. This was merged before, but some of your changes is chrashing ESP32-2432S028R board

I don't have access to this board. If you need assistance, feel free to share the logs. 
PS: Does it crash, if you try to access the config page?

**Comment by nitroxgas** [comment]
Hi Elmo128! The board crashes because there's no more heap space available to deal with the screens, and sometimes even https connections can't be established.
Some boards are dealing with extremely low resources, so my suggestion is that changes like implementing new features that are not shared among all boards should be protected by compilation directives (#ifdef).
Or even contained inside the driver area of that specific board. For example, I´ve implemented the touch function inside the board code to restrict him.

**Comment by elmo128** [comment]
> Some boards are dealing with extremely low resources, so my suggestion is that changes like implementing new features that are not shared among all boards should be protected by compilation directives (#ifdef).

> Or even contained inside the driver area of that specific board. For example, I´ve implemented the touch function inside the board code to restrict him.

Hi.
Really? Then the tft handling might need some changes, especially since there is no need to buffer a whole screen in ram when using pannels with controller and framebuffer. It might require some additional cpu time, but as significant portions of the screen should not change too often,...  By the way, afaik dynamic memory allocation on embedded devices should stay an exception and the exception handling in monitor.cpp might have a negative performance impact as well.

**Comment by nitroxgas** [comment]
I know! Exactly what I did with Sprites handling in the first commit. Creating and destroying them to release memory to the httpclient. Before that, httpclient not only couldn't connect, but it started to keep a part of the memory blocked, and the board became unresponsive and restarted. So, catching these exceptions was needed to prevent the complete crash over a few minutes. Now I'm slicing each screen into 4 to be able to deal with the small amount of memory available in the esp-wroom-32.

File: src/drivers/storage/storage.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+#define JSON_KEY_STATS2NV	"saveStats"
+
+struct TSettings
+{
+	char WifiSSID[80]{ DEFAULT_SSID };
+	char WifiPW[80]{ DEFAULT_WIFIPW };
+	char PoolAddress[80]{ DEFAULT_POOLURL };
+	char BtcWallet[80]{ DEFAULT_WALLETID };
+	uint32_t PoolPort{ DEFAULT_POOLPORT };
+	uint32_t Timezone{ DEFAULT_TIMEZONE };
```
Comment (nitroxgas) [inline src/drivers/storage/storage.h:31]: Please consider changing the type to suit timezones less than zero. (-3 Brazil)

Comment (elmo128) [reply src/drivers/storage/storage.h:31]: Oh thanks. How could i miss that? :)

Comment (BitMaker-hub) [reply src/drivers/storage/storage.h:31]: this is just for me to remember, jeje.

And again, thanks for your contribution

  json["poolString"] = poolString;
  json["portNumber"] = portNumber;
  json["btcString"] = btcString;
  json["gmtZone"] = GMTzone;
  json["saveStatsToNVS"] = String(saveStatsToNVS);

Comment (elmo128) [reply src/drivers/storage/storage.h:31]: > ok, Elmo, I'll have to make a PR to bring back same used names to not have to reedit all config from active Nerdminers. Even I like this new names, this a real problem for those who update.

Hi.
Sure, np, do whatever is neccessary.

**Pull request merged**